Import/Export Code

tfpk opened the pull request:
Added import/export code. 

Yet TODO:
- [x] Wording/Icon Changes
- [X] Code Refactoring? (mainly just reordering things)
- [X] File Uploads
- [X] Code for closing the modals could be better (especially when uploading)

tfpk committed:
client: add bot import and export

Signed-off-by: Tom Kunc
[ @cyphar: rebase ]
Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

tfpk commented:
Downloads:
![image](https://user-images.githubusercontent.com/10906982/103260160-c752c080-49f0-11eb-9a63-f05fa7d1d432.png)

Uploads:
![image](https://user-images.githubusercontent.com/10906982/103260174-ddf91780-49f0-11eb-899c-c73dad1e9409.png)

tfpk commented:
New install window, with file upload button:
![image](https://user-images.githubusercontent.com/10906982/103284984-ecffba00-4a30-11eb-8ce0-39e0c46f89e2.png)

cyphar commented:
Since you're going to be busy with organising murder, do you mind if I carry this one?

tfpk commented:
Sure I'm happy for you to, but I was just waiting on feedback -- the only thing left should be wording changes and I'm more than happy for someone else to do those if you'd prefer.

cyphar commented:
Sure. The two main changes (aside from this one) are to give system errors a different skin of some kind and to make the timestamps of messages actually tell you which day the message was sent. I just merged a few minor changes (#45) which includes local storage of basic settings (is the panel open? what is your name? is text-to-speech enabled?).

cyphar commented:
Here's my quick feedback after playing with it for a little bit:
* Upload and download seem like reasonable names.
  - But the modal titles need to be updated.
* The import modal doesn't appear to use the same monospace font as the export modal (at least on my Linux laptop -- your screenshots appear to use the same font?).
* I think the "exists in import" state should look a bit nicer than having the red-highlighted text "(exists in import)". Maybe the row should be tinted red or something like that (or maybe have another table which says "exists in both" -- or maybe even just auto-uncheck the box). 
* There should be "select all" and "deselect all" buttons for both import tables.
* The import tables feel a little bit cramped, but I'm not sure if there's really a way to make it any prettier.
* It seems like the download/upload buttons are a bit weirdly spaced in relation to the icons (a bit cramped)? Also maybe the icons should be standard download and upload icons?
* Now that there is `--accent-3` when rebasing make sure the button colours are the same for "cancel import" and "reset room".

cyphar commented:
But yeah, you will be very busy with murder tomorrow and Sunday so I will take this one off your plate. Thanks so much for working on this though, it makes my life a lot simpler. :relieved:

cyphar commented:
I've also noticed that bot auto-saving appears to not work with this PR. I will fix that too.

cyphar committed:
client: clear upload bot modal on organic close

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

cyphar committed:
client: fix new bot addition to database

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

cyphar committed:
client: importexport: basic UX cleanups

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

cyphar commented:
I made a fair number of minor changes, one of the big ones is that the input fields now disappear after you click "import" and the way that identical bots are handled is different. I think this is okay enough to be merged (assuming you think it looks reasonable @tfpk) and we can work on any additional cleanups later.

![2021-01-02-234821_1055x652_scrot](https://user-images.githubusercontent.com/2888411/103457645-1aac7200-4d55-11eb-80e2-90162847dbf9.png)

cyphar commented:
(I will add select all and deselect all in some unspecified future time if I decide we actually need it.)

cyphar committed:
client: importexport: handle identical bots differently

In order to avoid "churning" the bots by deleting them constantly,
identical bots are now simply retained and the user can choose if they
want to keep them. This also makes the UX a bit cleaner.

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

tfpk commented on client/index.html:
```diff
@@ -317,24 +329,24 @@ <h5>Already Installed Bots</h5>
             </table>
 
-            <h5>Importing Bots</h5>
+            <h5>Imported Bots</h5>
```
This implies the bot has already been imported, hence why I think I used "Importing". Maybe "New Bots" or "Bots to Import"?

cyphar replied:
New bots is probably better.

cyphar replied:
Done.

tfpk reviewed (commented):
Ah yeah I realised this was crappy but didn't have the wear with all to fix it...

Thanks :)

tfpk commented:
LGTM, nice work, and thanks for getting this over the line.

cyphar commented:
:shrug: I'll merge it now YOLO.
